### PR TITLE
Util changes needed for VSD

### PR DESCRIPTION
**Pull request by thk123**
Rebased version of #5205, with two additions:

 - tests for the const iterators added to base symbol table
 - stop definitions in symbol_table hiding the implemention in symbol_table_baset (risks these defintions going out of sync) 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: unit/util/irep.cpp
Comment (hannes-steffenhagen-diffblue) [inline unit/util/irep.cpp:25]: Side note: I’d really love if we could just get rid of `USE_STD_STRING` already. I’m not entirely convinced we keep it around for anything other than nostalgia at this point.

File: src/util/symbol_table_base.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/util/symbol_table_base.cpp:65]: 96bd3ae8e86ac8712a662e80d710c05a64b8442c `s/interate/iterate` in the commit message.

Also this commit should be removed; Current develop already has begin/end for `symbol_tablet`.

File: unit/util/symbol_table.cpp
Comment (hannes-steffenhagen-diffblue) [inline unit/util/symbol_table.cpp:20]: This commit (a9afb9a25f3fd32e5e227bc9a91d7d5228bf38f6) can also be removed; symbol table iterators are tested in `unit/util/symbol_table.cpp`

File: src/util/symbol_table.h
Comment (hannes-steffenhagen-diffblue) [inline src/util/symbol_table.h:124]: develop `symbol_table_baset::begin/end` are both purely virtual, so can not be used in this way.

**Review by hannes-steffenhagen-diffblue — CHANGES_REQUESTED** [review]
Other than the `USE_STD_STRING` change I don’t believe any of these changes should be merged as they already exist in some form or another on develop.

File: unit/util/symbol_table.cpp
Comment (thk123) [reply unit/util/symbol_table.cpp:20]: I can see a single class to `begin()` but no call to end. This also tests the correct dispatching when you have a `symbol_tablet` but the compile time type is `symbol_table_baset`

File: src/util/symbol_table.h
Comment (thk123) [reply src/util/symbol_table.h:124]: Discussed out of band: more consistent to have both const and non-const iterators be virtual and defined in the base class.

**Comment by thk123** [comment]
@hannes-steffenhagen-diffblue I've made the begin/end virtual in the base class as agreed, ready for [re-review](https://github.com/diffblue/cbmc/compare/acc46d626ecb385499286a86194eaa1527d1c09f..dca07dda7bb9a504cef9aca5d302d6f542f4267e)

File: unit/util/irep.cpp
Comment (martin-cs) [reply unit/util/irep.cpp:25]: I believe the only use case for it is / was debugging problems in the `dtringt` implementation.  I am not sure how many issues it will find that have not already been found.

**Review by martin-cs — APPROVED** [review]
Am happy once @hannes-steffenhagen-diffblue 's comments have been addressed.  I think that it will still work with the patch set.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
I still think it’s a bit inconsistent we have a pure virtual mut `begin/end` but an implementation for const `begin/end`. But I’d not block on that if that’s the only complaint.

**Comment by thk123** [comment]
@peterschrammel if you get a min, would you be able to provide the code owner review for this short PR? You reviewed this originally in #5205, Martin addressed your comments (this is a rebased version)

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
